plugin/pkg/auth/authenticator/token/openid/openid.go

```diff
@@ -0,0 +1,156 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// openid implements the authenticator.Token interface using the OpenID protocol.
+package openid
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/auth/user"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	"github.com/coreos/go-oidc/jose"
+	"github.com/coreos/go-oidc/oidc"
+	"github.com/golang/glog"
+)
+
+var (
+	maxRetries   = 5
+	retryBackoff = time.Second * 3
+)
+
+var errNoKeyEndpoint = fmt.Errorf("OpenID provider must provide 'jwks_uri' for public key discovery")
+
+type OpenIDAuthenticator struct {
+	client *oidc.Client
+}
+
+// NewOpenID creates a new OpenID client with the given clientID and discovery URL.
+// NOTE(yifan): For now we assume the server provides the "jwks_uri" so we don't
+// need to manager the key sets by ourselves.
+func NewOpenID(clientID, discovery, caFile string) (*OpenIDAuthenticator, error) {
+	var cfg oidc.ProviderConfig
+	var err error
+	var roots *x509.CertPool
+
+	url, err := url.Parse(discovery)
+	if err != nil {
+		return nil, err
+	}
+
+	if url.Scheme == "http" {
+		glog.Warning("Using insecure connection to the OpenID service")
```

Comment:
@yifan-gu thanks for the reponse.

Here are a few arguments I hope to convince you with.

1. Allowing http is actually a spec violation. See section 3. of https://openid.net/specs/openid-connect-discovery-1_0.html
> issuer
> REQUIRED. URL using the ***https*** scheme with no query or fragment component that the OP asserts as its Issuer Identifier. If Issuer discovery is supported (see Section 2), this value MUST be identical to the issuer value returned by WebFinger. This also MUST be identical to the iss Claim value in ID Tokens issued from this Issuer.

2. I don't think that the protection of a firewalled network is something that you want to rely on. It means that anybody that has access to your network can MITM this http exchange. With HTTPS, an attacker would need to have access to the keys on the machine which significantly reduces the surface of attack.

3. If customers really want this "feature", I think that you should at the very least protect it with a flag or some kind of global config so that the said customers avoid shooting themselves in the foot.

What do you think?